src/ptychography40/reconstruction/wdd/wdd_udf.py

```diff
@@ -204,7 +223,275 @@ def get_results(self):
             (self.results.cut)
             ).astype(self.params.complex_dtype)
         real_cut = real_cut/np.max(np.abs(real_cut))
+        res = real_cut.conj()
         return {
             'cut': self.results.cut,
-            'reconstructed': real_cut.conj(),
+            'reconstructed': res,
+            # for live plotting
+            'amplitude': np.abs(res),
+            'phase': np.angle(res),
+        }
+
+
+class PatchIndex(NamedTuple):
+    patch_index: int
+    within_patch: int
+
+
+class GridSpec:
+    def __init__(self, size: int, step: int):
+        if not isinstance(size, int):
+            raise ValueError(f"Size {size} must be integer")
+        if not isinstance(step, int):
+            raise ValueError(f"Step {step} must be integer")
+        if size < step:
+            raise ValueError(f"Size {size} < step {step}, has gap!")
+        if size <= 0:
+            raise ValueError(f"Size {size} must be positive")
+        if step <= 0:
+            raise ValueError(f"Step {step} must be positive")
+        self.size = size
+        self.step = step
+
+    def get_patch_indices(self, coord) -> List[PatchIndex]:
+        res = []
+        patch_index = coord // self.step
+        while patch_index * self.step + self.size > coord:
+            within_patch = coord - patch_index * self.step
+            res.append(PatchIndex(patch_index=patch_index, within_patch=within_patch))
+            patch_index -= 1
+        return res
+
+    @property
+    def min_patch_index(self):
+        candidates = self.get_patch_indices(0)
+        assert candidates
+        min_index = np.iinfo(np.dtype(np.int32)).max
+        for c in candidates:
+            min_index = np.minimum(min_index, c.patch_index)
+        return min_index
+
+    def max_patch_index(self, size):
+        return (size - 1) // self.step
+
+
+class PatchWDDUDF(UDF):
+    """
+    Reconstruct many small patches with WDD and merge them in the end
```

Comment:
For me, it was a bit confusing that `patch_trim` and `patch_overlap` are separate from `patch_shape`, even if I can understand the reason from the implementation side. Would it make sense to move all patch-related parameters to `recon_parameters` maybe?

Also, somewhere there should be documentation on how this all fits together - maybe a small diagram here using https://asciiflow.com or similar? Something like this but in complete/good:

![image](https://github.com/user-attachments/assets/18efc349-8e45-4582-af9d-9099657b5f8c)